README.md
# MotorCortex-2dcam

**Table of Contents**

- [MotorCortex-2dcam](#motorcortex-2dcam)
  - [Demo](#demo)
- [Intro / Features](#intro--features)
- [Getting Started](#getting-started)
  - [Installation](#installation)
  - [Importing and Loading](#importing-and-loading)
- [Creating Incidents](#creating-incidents)
  - [ZoomTo](#zoomto)
  - [FollowPath](#followpath)
- [Adding Incidents in your clip](#adding-incidents-in-your-clip)
- [Contributing](#contributing)
- [License](#license)
- [Sponsored by](#sponsored-by)

## Demo

[Check it out here](https://donkeyclip.github.io/motorcortex-2dcam/demo/)

# Intro / Features

MotorCortex-2dcam acts exactly as a 2d camera that can focus and zoom at
any given point of any element.
You can move your camera around, zoom in and out, follow paths, all
via its easy to use Incidents.

This Plugin exposes two Incidents:
- ZoomTo
- FollowPath

# Getting Started

## Installation

```bash
$ npm install --save @donkeyclip/motorcortex-2dcam
# OR
$ yarn add @donkeyclip/motorcortex-2dcam
```

## Importing and loading

```javascript
import { loadPlugin } from "@donkeyclip/motorcortex";
import TDCAMDef from "@donkeyclip/motorcortex-2dcam";
const TDCAM = loadPlugin(TDCAMDef);
```

# Creating Incidents

## ZoomTo

Any element can be selected by ZoomTo Incident and the camera will zoom
to any of its points, on any duration and by any easing.
Our viewport is always considered the parent of our target element. The
virtual camera will zoom to the given point of our target and will bring it in the middle
of our viewport, in the middle of its parent element, zoomed as per attributes passed.

```javascript
const zoomto = new TDCAM.ZoomTo(
  {
    animatedAttrs: {
      position: {
        x: 1280, // pixel 1280
        y: 150, // pixel 150
        zoom: 1.2, // zoom level, from 0 to inf., 1:1 with scale
      },
    },
  },
  {
    selector: ".img",
    duration: 4000,
    easing: "easeInOutSine",
  }
);
```

#### IMPORTANT

The element that you want to zoom to (any element that you target via your selector) must have `transform-origin: top left`

### ZoomTo Attrs
The expected attributes of ZoomTo include the animatedAttrs. This should contain the animated attribute `position`.
Position is a composite attribute consisting of x, y and zoom, where x and y are the coordinates of our selected element we want to focus and zoom the zoom level we want to apply.
All, x, y and zoom are optional and if not provided will not be affected during animation.

## FollowPath

Instead of poviding the point we want to focus on and let the camera animate its position
to it, we can provide (svg) paths, in the form of path definition ([https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d]). 
For this instead of using ZoomTo and providing x, y and zoom to define the target position of the camera,
you can use FollowPath and just provide a path that you want your camera to follow and your final zoom
level.

```javascript
const fpath = new TDCAM.FollowPath(
  {
    animatedAttrs: {
      position: {
        path: "M 1280 150 L 1380 464",
        zoom: 1.2,
      },
    },
    from: 0.2,
    to: 0.9,
    transition: 1000,
  },
  {
    duration: 3000,
    selector: ".img",
    easing: "easeOutCubic",
  }
);
```
### FollowPath Attrs
As you can see in the example, on our animatedAttrs we again used the position attribute but this
time we passed instead of x and y the `path` which we want our camera to follow.

If you want your camera to move only on a part of a path you can use the `from` and `to` attributes
that define the fraction (0 to 1) that you want your start and end point to be. E.g. if you
provide the values `from: 0.2` and `to: 0.9` then the camera will start from the 20% of the path's overall length and
will animate all the way to the 90% of its length. Of course, both start and end are optional and have
default values 0 and 1 respectively.

As we are moving our camera along a path, at the 0 millisecond of our Incident our camera will be placed on
the very first point of the provided path. This can cause a "jump" effect as the camera will
move to point 0 without animation. In order to handle this the `transition` attribute
is provided. If provided, transition defines (in milliseconds) an optional transition time from the current
camera position to point 0 of our path, so we avoid the jump. If provided then the camera will linearly move from
the current camera position to the point 0 position in the given milliseconds. The total duration of the
Incident will not be affected as the movement on the actual path will be
less by "transition" milliseconds. "transition" of course is optional with default value = 0.

# Adding Incidents in your clip

```javascript
clipName.addIncident(incidentName,startTime);
```

# Contributing 

In general, we follow the "fork-and-pull" Git workflow, so if you want to submit patches and additions you should follow the next steps:
1.	**Fork** the repo on GitHub
2.	**Clone** the project to your own machine
3.	**Commit** changes to your own branch
4.	**Push** your work back up to your fork
5.	Submit a **Pull request** so that we can review your changes


# License

[MIT License](https://opensource.org/licenses/MIT)

# Sponsored by
[<img src="https://presskit.donkeyclip.com/logos/donkey%20clip%20logo.svg" width=250></img>](https://donkeyclip.com)


CHANGELOG.md
# Changelog

### [0.3.1](https://github.com/donkeyclip/motorcortex-2dcam/compare/v0.3.0...v0.3.1) (2022-06-24)


### Bug Fixes

* import only effect instead of full MC ([f553415](https://github.com/donkeyclip/motorcortex-2dcam/commit/f5534153955eef5bee358c8f56fc05bbf5f654f5))

## [0.3.0](https://github.com/donkeyclip/motorcortex-2dcam/compare/v0.2.0...v0.3.0) (2022-04-07)


### Features

* update compatibility with motorcortex ([f97769b](https://github.com/donkeyclip/motorcortex-2dcam/commit/f97769bac17d9783da48c40c1dc4666aa9dc8244))

## [0.2.0](https://github.com/donkeyclip/motorcortex-2dcam/compare/v0.1.0...v0.2.0) (2022-04-07)


### Features

* Drop Explorer 11 support ([18a8621](https://github.com/donkeyclip/motorcortex-2dcam/commit/18a862163b1ff122aee9687f9330a370ec0c944b))

## [0.1.0](https://www.github.com/donkeyclip/motorcortex-2dcam/compare/v0.0.16...v0.1.0) (2022-03-17)


### Features

* declare mc 8 as compatible ([9438fe6](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/9438fe6f7f340c4d6f3e0960e6fe5f92090b38b2))


### Bug Fixes

* drop code from demo config ([aa7a4af](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/aa7a4af79b82d5af10686d00d4f22fc2e7c3310d))
* fix broken rollup config ([38e62e6](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/38e62e65ab0504e7bedee26e9e9b9f7b2094141c))
* fix broken webpack config ([d384ae8](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/d384ae8cbb4353481a1c84397a1af78de356b5ea))
* Typo fixes by misspell-fixer ([#9](https://www.github.com/donkeyclip/motorcortex-2dcam/issues/9)) ([1444780](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/144478048ab2cea39629124a4e6166e2492839aa))
* update build and demo files drop old babel config ([449793a](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/449793aa30e584c8d4b90185d2e2a018110b3392))
* update build and demo files drop old babel config ([70a35c2](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/70a35c2377c19364cb5ac80aba76e59e395aa897))
* update build and demo files drop old babel config ([0bf9796](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/0bf979620e4970c4f840ed6250a096328c1d2180))
* update build and demo files drop old babel config ([ed74044](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/ed740445217481a39e827b38584797e22dda994e))
* update build and demo files drop old babel config ([2bbfd9c](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/2bbfd9c4458aa8d823ac5cc804cf8d02de82370e))
* update build and demo files drop old babel config ([e085c0c](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/e085c0ce029c9d478d88ef11e4566c962d2cc750))
* update build and demo files drop old babel config ([a4d2125](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/a4d212584e1ae2ad47426110281cc357a02efb95))
* update build and demo files drop old babel config ([aaa8ad1](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/aaa8ad111fdef140b066d3698d446c096f9d63f3))
* update rollup deps to latest packages ([d04df95](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/d04df9502926a1a60c15c0503c48a716deec41ea))
* use esm instead of umd for demo page ([11a5fd7](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/11a5fd7390ea6b4f237b329ee4eb9cdcb29a744c))

### [0.0.16](https://www.github.com/donkeyclip/motorcortex-2dcam/compare/v0.0.15...v0.0.16) (2021-09-30)


### Bug Fixes

* update deps ([c7f1726](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/c7f17261f56ee2877cd452ebfb692c4ccac8906a))

### [0.0.15](https://www.github.com/donkeyclip/motorcortex-2dcam/compare/v0.0.14...v0.0.15) (2021-09-24)


### Bug Fixes

* update npm name ([369c9b4](https://www.github.com/donkeyclip/motorcortex-2dcam/commit/369c9b48a8c2b2fdcc85d1198225954064557825))

### 0.0.14 (2021-04-16)


### Bug Fixes

* add semantic release ([1ba2e9b](https://www.github.com/kissmybutton/motorcortex-2dcam/commit/1ba2e9bd512422622fecf2a79cdf50d8a66c6c1e))


rollup.config.js
import babel from "@rollup/plugin-babel";
import commonjs from "@rollup/plugin-commonjs";
import json from "@rollup/plugin-json";
import resolve from "@rollup/plugin-node-resolve";
import { terser } from "rollup-plugin-terser";
import pkg from "./package.json";
export default [
  {
    input: "src/index.js",
    external: ["@donkeyclip/motorcortex"],
    output: [
      { file: pkg.main, format: "cjs" },
      { file: pkg.module, format: "es" },
    ],
    plugins: [json(), resolve(), babel(), commonjs(), terser()],
  },
  {
    input: "src/index.js",
    external: ["@donkeyclip/motorcortex"],
    output: [
      {
        globals: {
          "@donkeyclip/motorcortex": "MotorCortex",
        },
        name: pkg.name,
        file: pkg.browser,
        format: "umd",
      },
    ],
    plugins: [
      resolve({ mainFields: ["module", "main", "browser"] }),
      babel(),
      commonjs(),
      terser(),
      json(),
    ],
  },
];


src/Incidents/ZoomTo.js
import { Effect } from "@donkeyclip/motorcortex";
import Adaptor from "../utils/Adaptor";

/**
 * Thus, here you'll find:
 * the following properties:
 * - this.element: provides a reference to the specific element of the MonoIncident
 * - this.attributeKey: the key of the animatedAttr of the MonoIncident
 * - this.targetValue: the final value of the animatedAttr
 * and the following methods:
 * - onGetContext
 * - getScratchValue
 * - onProgress
 * which are analysed more inline
 *
 **/
export default class MyEffect extends Effect {
  /**
   * the scratch value of the Incident should return back the triplette
   * x, y, zoom
   * We consider as the viewport the parent node of our element and we calculate
   * its initial position taking in consideration the relative position of our
   * element into its parent node. The initial zoom is calculated out of the
   * scaleX value of our element
   **/
  getScratchValue() {
    this.adaptor = new Adaptor(this.element);
    return this.adaptor.calcXYZoom();
  }

  onGetContext() {
    this.adaptor = new Adaptor(this.element);
    this.progressMethod = this.adaptor.createProgressFunction({
      start: this.initialValue,
      target: this.targetValue,
    });
  }

  onProgress(m) {
    const vals = this.progressMethod(this.getFraction(m));
    this.element.style.transform = `translateX(${vals.translateX}px) translateY(${vals.translateY}px) scaleX(${vals.scale}) scaleY(${vals.scale})`;
  }
}


demo/index.js
import MotorCortex from "@donkeyclip/motorcortex";
import Player from "@donkeyclip/motorcortex-player";
import MyPluginDefinition from "../dist/motorcortex-2dcam.esm";

const MyPlugin = MotorCortex.loadPlugin(MyPluginDefinition);

const rootClip = new MotorCortex.HTMLClip({
  html: `
    <div class="container">
      <div class="casi-container"></div>
    </div>`,
  css: `
    .container, .casi-container{
      width: 1920px;
      height: 1095px;
      background: black;
    }
  `,
  host: document.getElementById("clip"),
  containerParams: {
    width: "1920px",
    height: "1095px",
  },
});

const mypath = "M 500 250 L 960 547.6";
const clip = new MotorCortex.HTMLClip({
  html: `
    <div class="container">
      <div class="img">
        <img src="https://donkeyclip.github.io/motorcortex-slides/demo/assets/bg2.jpg" />
      </div>
    </div>`,
  css: `
    .container{
        width: 1920px;
        height: 1095px;
        background: black;
    }
    .img{
        width: 1920px;
        height: 1095px;
        transform-origin: top left;
    }
  `,
  selector: ".casi-container",
  containerParams: {
    width: "1920px",
    height: "1095px",
  },
});

rootClip.addIncident(clip, 0);

const zoomTo1 = new MyPlugin.ZoomTo(
  {
    animatedAttrs: {
      position: {
        x: 1075, 
        y: 678, 
        zoom: 1.2,
      },
    },
    initialValues:{
      position: {
        x: 1920/2, 
        y: 1095/2, 
        zoom: 1,
      },
    }
  },
  {
    selector: ".img",
    duration: 2000,
    easing: "easeOutSine",
  }
);

const zoomTo2 = new MyPlugin.ZoomTo(
  {
    animatedAttrs: {
      position: {
        x: 500, 
        y: 250, 
        zoom: 2,
      },
    },
  },
  {
    selector: ".img",
    duration: 4000,
    easing: "easeInOutSine",
  }
);


const pop1 = new MyPlugin.FollowPath(
  {
    animatedAttrs: {
      position: {
        path: mypath,
        zoom: 1,
      },
    },
  },
  {
    duration: 3000,
    selector: ".img",
    easing: "easeOutCubic",
  }
);

clip.addIncident(zoomTo1, 0);
clip.addIncident(zoomTo2, 2000);
clip.addIncident(pop1, 6000);

new Player({ clip: rootClip });
